### PR TITLE
feat(ts-sdk, vault): reach the refund from a reconstruction

### LINKED ISSUE #2203: ETH reorg recovery for an orphaned Pre-PegIn registration
Split out of #1463, which stays as the decision record and reasoning.

## The scenario

Two things must both happen, which is why this is a fail-safe rather than a feature:

1. An ETH reorg deeper than 8 blocks — never observed in Ethereum's history, and
2. The `submitPeginRequestBatch` transaction is **not** re-included afterwards — unusual, most reorged transactions return to the mempool and land within a block or two.

If both occur after the Pre-PegIn BTC has been broadcast, the on-chain vault record is gone, the indexer drops it too, and the depositor's BTC sits in an HTLC the app can no longer see.

## Agreed approach

Deliberately minimal — a documented manual path, not a built-out product surface.

1. A script scans the rollback window and identifies affected ETH addresses and their Pre-PegIn tx hashes.
2. A **hidden page behind a feature flag**, off by default, enabled only if this ever fires. The user connects their wallet, pastes their Pre-PegIn tx hash, and performs the refund.
3. The FE is given the affected addresses and prompts those users toward the page when their wallet connects.

Feasible because the depositor's secrets are re-derivable from their wallet plus the Pre-PegIn transaction.

## Actions

- [ ] **Protocol team** — provide the script that identifies affected ETH addresses and their Pre-PegIn tx hashes
- [ ] **Web team** — document the recovery mechanism in-repo, share with DG and Jenks for FAQ
- [ ] **Web team** — build the hidden feature-flagged page
- [ ] **Runbook + a practice run** (raised by @vitsalis) — a path nobody has rehearsed is not a path

## Priority

Low, pending bandwidth. Acceptable to defer **as long as the mechanism is documented** — the documentation and runbook matter more than the page, since a manual rescue is workable and an undocumented one is not.

## Context worth keeping

The indexer is Ponder-based and treats a block as final at **65 confirmations** (~13 min). An 8-confirmation gate (~1.6 min) leaves roughly 57 blocks where the UI has moved on but the indexer would still roll the record back. That gap is why this recovery path needs to exist.

### PR DESCRIPTION
**Pull request by gbarkhatov**
# Reach the refund from a reconstruction, and verify it against a real wallet

Follows [#2278](https://github.com/babylonlabs-io/babylon-toolkit/pull/2278). Closes [#2203](https://github.com/babylonlabs-io/babylon-toolkit/issues/2203).

PR 1 could verify a parameter set against a stranded Pre-PegIn. It could not get from there to a refund, and none of it had ever been run against a real wallet. This does both — and running it found a bug that two PRs of unit tests had missed.

## Verified end to end on signet

A full green run against a real UniSat wallet and a real signet vault, recovering blind and diffing against the row it never read:

```
Rehearsing against vault 0x7aca1953…c854 (status=expired, htlcVout=0)
  GROUND TRUTH (not fed into the reconstruction): hashlock=0x650945d2…694b, amount=1000000
  Wallet pubkey: 033f8f4496…c062
  ✔ derived hashlock MATCHES the on-chain hashlock — the key fits the lock.
  Pre-PegIn fee: Σin 1252998160 − Σout 1252997196 = 964 sat
  ✔ verified: 1 candidate(s) tried, terms projected.
  ✔ recovered peg-in amount MATCHES the on-chain amount.  (1000000)
  ✔ refund signed and signature verified against a recomputed sighash.
  ✔ would-be refund txid: fbc3960f…e856 (NOT broadcast)
```

Nothing was broadcast. The txid is computed from the signed transaction, never sent — though that vault is `expired`, so its timelock has matured and the transaction would be accepted.

And a second run against a **two-vault** Pre-PegIn, which exercises the sibling path — the reconstruction has to rebuild every HTLC output, and the refund path's anchor check is fail-closed, so a single-vault run leaves both untested:

```
Rehearsing against vault 0x00d643c4…d821 (status=depositor_withdrawn, htlcVout=1)
  Multi-vault Pre-PegIn: 2 siblings — the sibling path is exercised.
  GROUND TRUTH (not fed into the reconstruction): hashlock=0x0528d00c…f3a9, amount=2838383
  vaultCount=2, authAnchorHash=592c578e…6ec4
  ✔ derived hashlock MATCHES the on-chain hashlock — the key fits the lock.
  ✔ verified: 1 candidate(s) tried, terms projected.
  ✔ recovered peg-in amount MATCHES the on-chain amount.  (2838383)
  ✔ refund signed and signature verified against a recomputed sighash.
```

Both cases are run explicitly. `--txid=<prePeginTxid>` pins a deposit — which is also how a real recovery starts, from a hash the depositor supplies — and without it the rehearsal prefers a batched Pre-PegIn, since the sibling path is the riskier code. It states in its output which case it got, so a single-vault run cannot quietly look like full coverage.

Verified on signet: a single-vault `expired` deposit whose HTLC is unspent and matured, and two different two-vault deposits, at `htlcVout=0` and `htlcVout=1`.

Coverage is ranked above spendability on purpose: every batched Pre-PegIn this wallet has is `depositor_withdrawn`, so preferring a refundable deposit would skip the sibling path entirely. Nothing is broadcast, and a withdrawn deposit still proves derivation, verification, wallet signing and the recomputed-sighash check — so the run says out loud when its chosen HTLC is already spent rather than letting that pass silently.

This is the practice run the issue asked for and nobody had done.

## The bug it found

`deriveHashlocksFromPrePegin`, as merged in PR 1, could not be called with a real wallet:

```
Run failed: vaultContext: depositorBtcPubkey must be exactly 32 bytes, got 33
```

Wallets return the **33-byte compressed** key from `getPublicKey`; the function only stripped `0x` and required x-only. Every unit test passed because every one fed bare x-only hex — the same blind spot review had already flagged on the sibling function, which was fixed there and missed here.

Fixed with `processPublicKeyToXOnly`, plus regression tests asserting the compressed and `0x`-prefixed forms derive identical secrets to the x-only form.

## `toRefundInputs`, and no second refund implementation

`buildAndBroadcastRefund` takes its two reads — the vault row and the version-pinned Pre-PegIn context — as **injected callbacks** rather than performing them itself. That is exactly the seam recovery needs: a stranded deposit has no row to read, but it has a reconstruction that was byte-verified against the funded transaction, and those are the same fields the callbacks return.

```ts
const { vault, context } = toRefundInputs(result, { htlcVout, ... });

await buildAndBroadcastRefund({
  vaultId,
  readVault: async () => vault,
  readPrePeginContext: async () => context,
  feeRate, signPsbt, broadcastTx,
});
```

So the refund-fee cap, the abort checks, the `htlcVout` contiguity invariant, the signature verification and the bitcoind error classification stay in one place for both paths instead of being reimplemented for the rare one. The rehearsal runs that real orchestrator with a capturing broadcaster, which is why it exercises wallet signing under the taproot script-path options — CLAUDE.md critical path 8, where support is inconsistent and failures are silent.

`applicationEntryPoint` is the one field the reconstruction does not carry: not recoverable from the Bitcoin transaction, and a property of the vault-provider registry row, which is address-keyed and survives. It is an explicit input, not a silent default.

## The headless proof

A refund PSBT built from parameters recovered blind is asserted **byte-identical** to one built from the parameters the deposit was actually made with. The plan assumed this needed a wallet and a signet run; it does not, because `buildRefundPsbt` takes parameters directly and building a PSBT involves no signing. A mutation check sits beside it — perturbing `timelockRefund` must make the builder refuse — because a byte comparison that cannot fail proves nothing.

## Tests

46 in the recovery module, all against real WASM, including one pinning that a wallet's compressed pubkey survives the projection unchanged. `lint`, `tsc --noEmit` and the full suites pass across the SDK, the vault app and the e2e project: **1591** SDK tests, 3387 vault tests.

## Scope, honestly

**Not covered by the live runs:** reading the version stamps at a **historical block height** — tracked in [#2292](https://github.com/babylonlabs-io/babylon-toolkit/issues/2292). The rehearsal reads them at `latest`, which is correct for a current vault because current values are the stamped values — but a real incident where a version bumped or an operator rotated between deposit and recovery needs historical reads, and nothing in `clients/eth` accepts a block number today.

Worth knowing before merge: the verifier catches a wrong `timelockRefund` or participant set, but **not** a wrong `vaultCoreVersion` — v1 and v2 give byte-identical connector scriptPubKeys and the amount is inverted using the supplied reserve, so a wrong value yields a sole match with a shifted split and no error. That is why #2292 is a real gap rather than a tidy-up, and it is pinned by a deliberate negative test.

The two-vault run used a `depositor_withdrawn` deposit, so its HTLC output is already spent: the refund signs and verifies, but that particular transaction could not be broadcast. The single-vault run used an `expired` deposit whose output is unspent and whose timelock has matured, so that one could.

**Deliberately not here:** the forked-block scanning script (protocol side), the feature-flagged recovery page (agreed deferrable provided the mechanism is documented, which `RECOVERY.md` now does), and sharing the mechanism for the FAQ.

## Review

Feeds CLAUDE.md critical paths 1 and 8 — the projected values reach a signed Bitcoin transaction. `toRefundInputs` performs no validation of its own: every value it emits comes from a parameter set already byte-matched against the funded transaction's HTLC outputs, and the orchestrator re-checks what it owns.

### CODE REVIEW

File: services/vault/e2e/real/actions/recover.ts [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+      return await p[method](...args);
+    },
+    { providerPath, method, args },
+  );
+
+  let done = false;
+  const sweeper = (async () => {
+    while (!done) {
+      await ctx.page.waitForTimeout(400);
+      await sweepApprovals(ctx.context, ctx.page, ctx.log).catch(() => {});
```
[the remark this answers is not in the document]
Comment (gbarkhatov) [reply services/vault/e2e/real/actions/recover.ts]: Fixed. The sweeper now reports a failure instead of discarding it — you and the other bot both landed on this, and CLAUDE.md is explicit that an empty catch is not acceptable.

One deviation from the suggestion: reported **once per `callProvider` invocation** rather than per tick. The loop runs every 400ms against a 120s timeout, so a persistent fault would otherwise emit ~300 identical lines and bury the thing it is trying to surface.

File: packages/babylon-ts-sdk/src/tbv/core/recovery/RECOVERY.md [hunk 2]
```diff
@@ -36,7 +36,27 @@ Parameters from the recovery script are untrusted input and must never feed a si
...
+  vaultId,
+  readVault: async () => vault,
+  readPrePeginContext: async () => context,
+  feeRate, signPsbt, broadcastTx,
+});
+```
+
+There is deliberately no separate refund implementation for recovery: the fee cap, the abort checks, the `htlcVout` contiguity invariant and the bitcoind error classification stay in one place for both paths. `applicationEntryPoint` is the one field not carried by the reconstruction — take it from the vault-provider registry row, which is address-keyed and survives.
+
+A test asserts that a refund PSBT built from recovered parameters is **byte-identical** to one built from the parameters the deposit was actually made with. `broadcastTx` is an injected callback, so the whole path can be rehearsed with a capturing broadcaster that never touches the network.
```
[the remark this answers is not in the document]
Comment (gbarkhatov) [reply packages/babylon-ts-sdk/src/tbv/core/recovery/RECOVERY.md]: Fixed, and the second half mattered more than the first.

`RECOVERY.md` now names the concrete rehearsal — the `recover` action, with the command to run it — in the injected-broadcaster section. And "Before anyone needs this" no longer claims the path has never been rehearsed, because this PR rehearsed it: single-vault and two-vault, against a real UniSat wallet on signet.

It also records what the rehearsal still does NOT cover — reading the version stamps at a historical block height — so the correction does not overstate itself.

File: packages/babylon-ts-sdk/src/tbv/core/recovery/toRefundInputs.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    offchainParamsVersion: offchainParams.version,
+    appVaultKeepersVersion: participants.appVaultKeepersVersion,
+    universalChallengersVersion: participants.universalChallengersVersion,
+    vaultProvider: participants.vaultProvider,
+    applicationEntryPoint,
+    amount: peginAmounts[htlcVout],
+    unsignedPrePeginTxHex: fundedPrePeginTxHex,
+    depositorBtcPubkey,
+    batch,
+  };
```
[the remark this answers is not in the document]
Comment (gbarkhatov) [reply packages/babylon-ts-sdk/src/tbv/core/recovery/toRefundInputs.ts:145]: Not applying this one; the premise does not hold.

`toRefundInputs` never calls `buildRefundPsbt`. It produces `VaultRefundData`, whose contract is "32 **or 33** bytes of hex (optional 0x prefix)" — validated as such at `buildAndBroadcastRefund.ts:322` — and the orchestrator narrows it itself at :466 with `processPublicKeyToXOnly` before the key reaches the builder at :513. Its own comment says the key "may arrive as wallet-native compressed sec1".

So narrowing early would discard a parity byte the consumer deliberately accepts, for a builder this module does not invoke. Wallets return the compressed form — the signet run derived from `033f8f44…` — so the compressed path is the ordinary one, not an edge case.

Added a test pinning that a compressed key survives the projection unchanged, so the contract is enforced rather than merely documented.

File: services/vault/e2e/real/actions/recover.ts [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+    }
+  })();
```
[the remark this answers is not in the document]
Comment (gbarkhatov) [reply services/vault/e2e/real/actions/recover.ts:196]: Fixed, both halves.

`AbortSignal.timeout(ESPLORA_TIMEOUT_MS)` bounds the request, and the `vout` index is checked before use — a short array previously produced `BigInt(undefined)`, a `TypeError` naming neither the transaction nor the index. Correct on both counts.

File: services/vault/e2e/real/actions/recover.ts [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+  for (const input of tx.ins) {
+    const txid = Buffer.from(input.hash).reverse().toString("hex");
+    const res = await fetch(`${btcApi}/tx/${txid}`);
+    if (!res.ok)
+      throw new Error(
+        `Esplora ${res.status} pricing input ${txid}:${input.index}`,
+      );
+    const prev = (await res.json()) as { vout: { value: number }[] };
+    totalIn += BigInt(prev.vout[input.index].value);
+  }
```
[the remark this answers is not in the document]
Comment (gbarkhatov) [reply services/vault/e2e/real/actions/recover.ts:243]: Fixed — see the reply on the sibling thread. Reported once per `callProvider` call rather than per 400ms tick, so a persistent fault surfaces without flooding the log.

File: services/vault/e2e/real/actions/recover.ts [hunk 6]
```diff
@@ -0,0 +1,483 @@
...
+    // Compare in the same form: wallets return the 33-byte compressed key,
+    // the row stores x-only. Comparing raw would flag every healthy run.
+    const walletXOnly = stripHex(walletPubkey).slice(-64).toLowerCase();
+    if (walletXOnly !== stripHex(target.depositorBtcPubKey).toLowerCase())
+      log(
+        `  NOTE: the connected wallet's key (${walletXOnly}) is not the row's ` +
+          `depositorBtcPubKey (${stripHex(target.depositorBtcPubKey)}). Derivation ` +
+          `will fail below — the root is bound to the account and network, not ` +
+          `just the seed.`,
+      );
```
[the remark this answers is not in the document]
Comment (gbarkhatov) [reply services/vault/e2e/real/actions/recover.ts]: Fixed. The comparison now runs both keys through `processPublicKeyToXOnly` instead of `slice(-64)`.

You are right about the failure mode: for a 65-byte uncompressed key the last 64 hex chars are the Y coordinate, so the check would have logged a false mismatch while the derivation succeeded — a diagnostic that lies is worse than none. UniSat returns compressed so it was latent, but latent is not fixed.

File: services/vault/e2e/real/actions/recover.ts [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+    log(
+      `  ✔ derived hashlock MATCHES the on-chain hashlock — the key fits the lock.`,
+    );
+
+    // ---- Parameters. Supplied from the stamped versions rather than searched;
+    // the enumeration fallback is unit-tested and is not the unknown here.
+    const publicClient = createPublicClient({
+      chain: sepolia,
+      transport: http(ethRpcUrl),
+    });
```
[the remark this answers is not in the document]
Comment (gbarkhatov) [reply services/vault/e2e/real/actions/recover.ts:413]: Fixed the two that matter.

`"signet" as never` is gone: there is now a typed `Record<NetworkName, BtcNetwork>` map, with the type taken from the SDK's own signature rather than imported from the WASM package (not a direct dependency of this project). A cast to `never` on a value that reaches HTLC script construction defeats exactly the check that would catch a wrong network. And `10_000 - 1` is now `REHEARSAL_MAX_COMMISSION_BPS`, with a note on why the widest admissible bound is the right choice for a rehearsal.

Left the viem chain hardcoded to `sepolia`: both networks resolve to it, and a one-entry map would be ceremony rather than safety. Happy to revisit if a non-Sepolia network is ever added.

File: services/vault/e2e/real/actions/recover.ts [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+      const key = v.unsignedPrePeginTx as string;
+      byPrePegin.set(key, [...(byPrePegin.get(key) ?? []), v]);
+    }
+    const groups = [...byPrePegin.values()].sort((a, b) => {
+      if (a.length !== b.length) return b.length - a.length;
+      return Math.min(...a.map(statusRank)) - Math.min(...b.map(statusRank));
+    });
+    const group = groups[0];
+    const usable = [...group].sort((a, b) => statusRank(a) - statusRank(b));
+    const target = usable[0];
```
[the remark this answers is not in the document]
Comment (gbarkhatov) [reply services/vault/e2e/real/actions/recover.ts:330]: Not applying this one — it would remove the coverage this PR exists to add.

All five batched Pre-PegIns for the test wallet are `depositor_withdrawn`. Filtering groups to those containing a refundable vault would have made the rehearsal skip every multi-vault deposit and fall back to a single-vault one, so the sibling path — the reconstruction rebuilding **every** HTLC output, and the refund path's fail-closed anchor check — would have gone untested. That is the riskiest code here, and it is the reason batches are ranked first. The proposed `throw` would also make the action fail outright on a wallet whose deposits have all been withdrawn, which is a worse outcome for a diagnostic.

Ranking coverage above spendability is deliberate, and sound because **nothing is broadcast**: a withdrawn deposit still proves derivation, verification, wallet signing and the recomputed-sighash check. Both multi-vault runs did exactly that.

The real risk you are pointing at is selecting a spent output *silently*, and that is worth fixing. The action now says so explicitly when the chosen deposit is not refundable, and points at `--txid` to target one that is:

```
NOTE: status "depositor_withdrawn" — this HTLC output is already spent.
Fine for a rehearsal, but this refund could not be broadcast.
Pass --txid=<prePeginTxid> to target a refundable deposit instead.
```

**Comment by gbarkhatov** [comment]
Split the one gap this PR does not cover into [#2292](https://github.com/babylonlabs-io/babylon-toolkit/issues/2292), so it does not vanish when `Closes #2203` fires on merge.

Recovery reads protocol parameters at `latest`. Correct for a current vault — which is why both signet runs pass — and wrong for a deposit stranded long enough for a version bump or an operator rotation to land in between. The verifier catches a wrong `timelockRefund` or participant set, but not a wrong `vaultCoreVersion`: v1 and v2 give byte-identical connector scriptPubKeys and the amount is inverted using the supplied reserve, so it returns a sole match with a shifted split and no error.

Noted in the description's scope section too.

**Review by jonybur — APPROVED** [review]
The seam is right: `buildAndBroadcastRefund` already takes the vault row and the version-pinned Pre-PegIn context as injected callbacks, so recovery does not need a parallel refund implementation - and this PR resists the temptation to build one. `toRefundInputs` projects a verified reconstruction into the two shapes the orchestrator expects, keeping the fee cap, the anchor check, the htlcVout contiguity invariant and the bitcoind error classification in one place for both paths.

The safety property the whole reconstruction exists to provide - that a recovered refund PSBT equals the one the deposit was actually made with, byte for byte - is proven by the byte-identity test. Its mutation counterpart (perturbing `timelockRefund` breaks the taptree so the builder refuses) shows that comparison has teeth; without it, the identity test would pass for any reconstruction at all. The batch-ordering, htlcVout-in-range and hashlock/amount length checks belong to the projection - the deeper checks stay with the orchestrator.

`applicationEntryPoint` is deliberately supplied rather than reconstructed (address-keyed row, survives the reorg), and `numLocalChallengers` derives from the keeper set per the btc-vault graph.rs `derive_challengers` reference. `depositorBtcPubkey` is passed through in whatever form the wallet returns; the orchestrator's own validation accepts 32-or-33-byte hex and narrows with `processPublicKeyToXOnly`, so pre-narrowing would discard the parity byte a compressed key carries.

The `deriveHashlocksFromPrePegin` fix - accepting the 33-byte compressed key wallets actually return - closes a case every unit test missed by feeding x-only hex, caught by the signet rehearsal against a real UniSat wallet. That rehearsal is the standing story for this path: single-vault and two-vault runs on signet, signed and signature-verified against a recomputed sighash, never broadcast. The RECOVERY.md note about what the rehearsal still does not cover (reading stamps at a historical block height) is honest and correct.

The e2e `recover` action's sweeper failure is reported once per `callProvider` invocation rather than silently swallowed; Esplora requests are timeout-bounded and vout indices checked before use; the network descriptor is a typed Record instead of `as never`. All the earlier defects hold up to a re-read of the current diff.

Approving.

**Review by jeremy-babylonlabs — APPROVED** [review]

**Review by jrwbabylonlab — APPROVED** [review]
